File: 1-Easy/Retro Typewriter Art/retro-typewriter-art.ts
const abbreviations: { [key: string]: string } = {
  sp: " ",
  bS: "\\",
  sQ: "'",
  nl: "\n"
};

const recipe: string[] = readline().split(' ');
let image: string = "";
let n: number = 0;

for (const chunk of recipe) {
  // takes the last 2 letter of chunk, if it's not in abbreviations, then it's undefined
  const abbreviationChar: string | undefined = abbreviations[chunk.slice(-2)];
  if (abbreviationChar === "\n") {
      // nl doesn't have any number before it, so this part is to skip the for loop
      image += abbreviationChar;
  }
  else if (abbreviationChar) {
      n = parseInt(chunk.slice(0, -2)); // all characters in the string except the last 2
      for (let i = 0; i < n; i++) {
          image += abbreviationChar;
      }
  }
  else {
      // not an abbreviation, since abbreviationChar is undefined
      n = parseInt(chunk.slice(0, -1)); // all characters in the string except the last one
      for (let i = 0; i < n; i++) {
          image += chunk.slice(-1);
      }
  }
}

console.log(image);

File: 1-Easy/Retro Typewriter Art/retro-typewriter-art.cs
using System;
using System.Collections.Generic;

class Solution
{
    static void Main(string[] args)
    {
        Dictionary<string, char> abbreviations = new Dictionary<string, char>
        {
            { "sp", ' ' }, 
            { "bS", '\\' },
            { "sQ", '\'' },
            { "nl", '\n' }
        };

        string[] recipe = Console.ReadLine().Split(' ');
        string image = "";
        int n = 0;
        
        foreach (string chunk in recipe)
        {
            // check if abbreviations contains a key by last 2 characters of the string, if yes then return it or 0
            char abbreviationChar = abbreviations.GetValueOrDefault(chunk.Substring(chunk.Length - 2, 2));

            if (abbreviationChar == '\n')
            {
                // nl doesn't have any number before it, so this part is to skip the for loop
                image += abbreviationChar;
            }
            else if (abbreviationChar != 0)
            {
                // it's an abbreviation, since it's not the default char value
                n = Convert.ToInt32(chunk.Substring(0, chunk.Length - 2));  // all characters in the string except the last 2
                for (int i = 0; i < n; i++)
                {
                    image += abbreviationChar;
                }
            }
            else
            {
                n = Convert.ToInt32(chunk.Substring(0, chunk.Length - 1)); // all characters in the string except the last one
                for (int i = 0; i < n; i++)
                {
                    image += chunk[chunk.Length - 1];
                }
            }
        }

        Console.WriteLine(image);
    }
}